### DISCUSSION TITLE
Deltas Since Last Dump?

### DISCUSSION THREAD

**Discussion by levigroker** [rusq/slackdump#193] [Q&A] [answered]
I'd love to be able to run this on a regular basis, but doing so appears to download everything since the dawn of time...

Does the underlying service provide "last modified" date, or similar, such that one could only export things which have a newer modification date since the previous export? I don't see any fields in the channel data which look like this. I see a "created" field, but that's it.

Am I missing it, or does the underlying system (Slack) not have this information... any other ideas on how this could be achieved?

Cheers,

Levi

**Comment by rusq** [comment, marked as answer]
Hey @levigroker ,

For now, it does not provide this feature out of the box (it's planned for when the state storage is implemented).  In v2 the only way to do it is to store and use the timestamp externally.

For example:
```
1. let t = current time.
2. read previousT from external storage (it may be null, which would mean we need to get everything from the beginning of time)
3. run Slackdump for the time interval of [previousT, t), if using cli:

   ./slackdump -dump-from previousT -dump-to t <channels>

5. if completed successfully, store t to external storage.
```

**Reply by levigroker** [reply]
Thank you @rusq ! That's what I was missing. I didn't see that in the docs, but it is clearly in the command help, so sorry I missed it, but this is great and something I can work with. Much appreciated.

**Reply by levigroker** [reply]
Curious about the timestamp... it is documented as of the format `2023-03-01T10:25:30` but I'm seeing some discrepancies in what's coming back (missing messages), so I'm wondering if there's time zone considerations for this timestamp?

**Reply by rusq** [reply]
Interesting, — what happens is:
1. the date specified in the parameters is converted verbatim to the Unix timestamp [i.e. "Sat May  7 22:18:32 2022" becomes "1651925912.000000"](https://github.com/rusq/slackdump/blob/4a2f0580690727f4b87e47859153107a353994c7/internal/structures/slacktime_test.go#L52)
2. this value is sent to the API.
3. Data is returned by the API

I think it might be something to do with timezones, as you suggested, I'll test that tonight and let you know.

Could you give more details on which messages are being missed?

For example, if you're running 2 times, first: from $T_1$ to $T_2$ and second: from $T_2$ to $T_3$, are you able to find $t_1\in[T_1,T_2]$ and $t_2\in[T_2,T_3]$, such as if you run it third time for the interval from $t_1$ to $t_2$ you can see those missed messages?  I'm interested in the value of $T_2$ and the timestamp of the message that was missed

**Reply by levigroker** [reply]
Hi, I'm not sure I follow exactly what you're asking for, but I think it's as simple as knowing the times here are expected to be in UTC. So, using the `-u` flag to `date` (like `CURRENT_DATE="$(date -u "+%Y-%m-%dT%H:%M:%S")"` gets the time in the proper format (per `slackdump`'s docs) and in the correct time zone (UTC).

**Reply by levigroker** [reply]
I was able to make this change and now get the messages I expect, based on the times I provide to the tool.

**Reply by levigroker** [reply]
FYI, I have developed this script as a helper to `slackdump` https://gist.github.com/levigroker/fa7b231373e68269843aeeee5cc845a3

**Reply by rusq** [reply]
Hey @levigroker,  that's amazing!  Thank you for the update, sorry for not explaining it properly.

Would you not object if I add this script to the `contrib` folder of slackdump?  This is the place for auxiliary tools that can make life easier for other users, I think other users of Slackdump might find it useful too.

PS - you can always make changes to it by submitting the PR, if you ever need to do that in future.

**Reply by levigroker** [reply]
@rusq Thanks! No worries on the time thing... it was just easier for me to test my assumption it was wanting UTC.
I think it would be great if you want to add my script to the `contrib` folder for others to use. I hope people find it helpful. Cheers!

**Reply by rusq** [reply]
Thank you for your contribution: https://github.com/rusq/slackdump/pull/198

**Reply by paravz** [reply]
Thanks for the tool, the script and this thread!

Helped me to create a hacky variant of the "backup everything since" invocation:
`./slackdump -r json -download -export ./slack-export -dump-from $(date -Iseconds --utc -r ./slack-export/users.json| perl -pe "s|\+00:00||;")`

PS: assumes dates are stored in utc on slack side

**Comment by rusq** [comment]
Hey @levigroker, thought maybe you'd like to try the new `resume` function (currently in master, but will be in v3.1) :)  I'm keen to hear some feedback on the feature - i.e. maybe something is missing etc.  The only caveat - it only supports the archive mode which stores data in the sqlite database, but can be converted at will to slack export (and from slack-export back to the database format, if required).
